### ISSUE TITLE
User extensions: empty collections and null

### ISSUE THREAD

**Issue by scurest** [KhronosGroup/glTF-Blender-IO#870]
Any key-value pair whose value is `{}`, `[]`, or `None` gets scrubbed out of the glTF data when exporting. That means a user-extension can't do things like this:

````
    "extensions": { "KHR_materials_unlit": {} }
    # gets turned into "extensions": {}
````

The way the in-tree code works around this is to use a hard-coded whitelist of key-names that shouldn't be scrubbed.

https://github.com/KhronosGroup/glTF-Blender-IO/blob/3ab37ba38a3ae483d69a029f979286ded8b9b94b/addons/io_scene_gltf2/blender/exp/gltf2_blender_export.py#L108-L115

That's not very robust and it doesn't help user extensions anyway. Here's my [proposed fix](https://github.com/KhronosGroup/glTF-Blender-IO/compare/master...scurest:empty-cols-in-json):

* Add some new objects you should use wherever you want an empty collection/null

    ````py
    EmptyDict = object()  # use this where you want an empty dict
    EmptyList = object()  # use this where you want an empty list
    Null = object()  # use this where you want a null
    ````
* Make `from_extension` believe that `EmptyDict` is a dict so it will allow it in the `"extensions"` object.
* Change `__fix_json` to convert `EmptyDict` into `{}` etc. Get rid of the whitelist.
* Whenever you want to put an empty dict, just use `EmptyDict` instead. Example:

    ````py
    extensions['KHR_materials_unlit'] = Extension(
        'KHR_materials_unlit',
        EmptyDict,
        required=False,
    )
    ````

What's everyone think?

cc @jjcasmar

**Comment by jjcasmar** [comment]
I agree. I have the same doubts when I first looks at the __fix_json. I think your solution is a good one.

**Comment by emackey** [comment]
Do 3rd-party addons have to import `EmptyDict` and friends?  What does that code look like?

I certainly agree that having a hard-coded list there is not acceptable now for 3rd-party extensions.

**Comment by scurest** [comment]
They just do it the same way they import everything else

````py
class glTF2ExportUserExtension:
    def __init__(self):
        import io_scene_gltf2
        self.Extension = io_scene_gltf2.io.com.gltf2_io_extensions.Extension
        self.EmptyDict = io_scene_gltf2.io.com.gltf2_io.EmptyDict
        
    def gather_material_hook(self, material, blender_material, export_settings):
        if material.extensions is None:
            material.extensions = {}
        material.extensions['KHR_materials_unlit'] = self.Extension(
            name='KHR_materials_unlit',
            extension=self.EmptyDict,
            required=False,
        )
````

**Comment by julienduroure** [comment]
Seems #1429 fixed that (at least for {}, didn't test [] or None)

**Comment by jjcasmar** [comment]
Actually, it only solves it for extras, no? Here we talk about empty dicts or lists for extensions. We talked about some solutions time ago but noone submit anything
